src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -617,6 +617,21 @@ private void VerifyAndUpdateWrite(int count)
             _responseBytesWritten += count;
         }
 
+        protected void VerifyResponseContentLength()
+        {
+            var responseHeaders = FrameResponseHeaders;
+
+            if (!HttpMethods.IsHead(Method) &&
```

Comment:
Should you check that if `HttpMethods.IsHead(Method) && responseHeaders.ContentLengthValue.HasValue` then `_responseBytesWritten == 0 || _responseBytesWritten == responseHeaders.ContentLengthValue.Value`?